server/src/main/java/org/opensearch/node/remotestore/RemoteStorePinnedTimestampService.java

```diff
@@ -0,0 +1,272 @@
+        Supplier<RepositoriesService> repositoriesService,
+        Settings settings,
+        ThreadPool threadPool,
+        ClusterService clusterService
+    ) {
+        this.repositoriesService = repositoriesService;
+        this.settings = settings;
+        this.threadPool = threadPool;
+        this.clusterService = clusterService;
+
+        pinnedTimestampsSchedulerInterval = CLUSTER_REMOTE_STORE_PINNED_TIMESTAMP_SCHEDULER_INTERVAL.get(settings);
+    }
+
+    /**
+     * Starts the RemoteStorePinnedTimestampService.
+     * This method validates the remote store configuration, initializes components,
+     * and starts the asynchronous update task.
+     */
+    public void start() {
+        validateRemoteStoreConfiguration();
+        initializeComponents();
+        startAsyncUpdateTask();
+    }
+
+    private void validateRemoteStoreConfiguration() {
+        final String remoteStoreRepo = settings.get(
+            Node.NODE_ATTRIBUTES.getKey() + RemoteStoreNodeAttribute.REMOTE_STORE_SEGMENT_REPOSITORY_NAME_ATTRIBUTE_KEY
+        );
+        assert remoteStoreRepo != null : "Remote Cluster State repository is not configured";
+        final Repository repository = repositoriesService.get().repository(remoteStoreRepo);
+        assert repository instanceof BlobStoreRepository : "Repository should be instance of BlobStoreRepository";
+        blobStoreRepository = (BlobStoreRepository) repository;
+    }
+
+    private void initializeComponents() {
+        String clusterName = ClusterName.CLUSTER_NAME_SETTING.get(settings).value();
+        blobStoreTransferService = new BlobStoreTransferService(blobStoreRepository.blobStore(), this.threadPool);
+        pinnedTimestampsBlobStore = new RemoteStorePinnedTimestampsBlobStore(
+            blobStoreTransferService,
+            blobStoreRepository,
+            clusterName,
+            this.threadPool,
+            ThreadPool.Names.REMOTE_STATE_READ
+        );
+    }
+
+    private void startAsyncUpdateTask() {
+        asyncUpdatePinnedTimestampTask = new AsyncUpdatePinnedTimestampTask(logger, threadPool, pinnedTimestampsSchedulerInterval, true);
+    }
+
+    /**
+     * Pins a timestamp in the remote store.
+     *
+     * @param timestamp The timestamp to be pinned
+     * @param pinningEntity The entity responsible for pinning the timestamp
+     * @param listener A listener to be notified when the pinning operation completes
+     * @throws IllegalArgumentException If the timestamp is less than the current time minus one second
+     */
+    public void pinTimestamp(long timestamp, String pinningEntity, ActionListener<Void> listener) {
+        // If a caller uses current system time to pin the timestamp, following check will almost always fail.
+        // So, we allow pinning timestamp in the past upto some buffer
+        if (timestamp < TimeUnit.NANOSECONDS.toMillis(System.nanoTime()) - TIMESTAMP_PINNING_PAST_BUFFER_IN_MILLIS) {
```

Comment:
10s seems too high? ideally consumers should not be using current system time and only rely on nanoTime. If consumers are using nanoTime, then isn't 10 sec buffer too much?

Reply:
This also depends on when the consumer is fetching the `timestamp`. For one such consumer, snapshot, it needs to get the latest timestamp, perform some operations and then pin the timestamp. Keeping this buffer only for that. As long as we make sure this buffer is considered in GC flow, we should be fine.

Reply:
Let me make this configurable.